### PR TITLE
*: Add module health_controller and move SlowScore, SlowTrend, HealthService from PdWorker to it

### LINKED ISSUE #16297: Performance/health feedback to client and the Unified Health Controller
## Development Task

Currently, there are several ways to evaluate whether a TiKV node is in healthy state. They are:

* SlowScore which is a way to evaluate whether raftstore is in a abnormally slow state. It can be informed to PD which enables PD to trigger the evict-slow-store-scheduler to avoid write requests to be sent to the store.
* SlowTrend, which is improved from SlowScore
* gRPC HealthService, which is a way built-in in gRPC for indicating whether the service is available to serve.

And all these things are managed in PdWorker.

We found the evict-slow-store-scheduler mentioned before is quite helpful to solve the problem that some TiKV nodes encountering slow IO might significantly affect the whole cluster's performance. However, there can still be problem if follower read is being used in the cluster, since it doesn't need to be processed on leader.

To solve the problem, we want the client-go to know whether each TiKV node is abnormal and avoid sending follower read requests to the problematic ones. We are now considering making it able to send some of the health information to the client via kv responses, so that the client has a more efficient and up-to-time information about the TiKV nodes' status and adjust the policy to select replicas.

We are planning to add a component named Unified Health Controller, which will be the unified entrance for managing and accessing the health status of the TiKV node. The SlowScore, SlowTrend and gRPC HealthService mentioned above should be moved to it, and PdWorker will still be responsible for updating them. The component itself should be outside PdWorker, which enables us to access it elsewhere, or add more information to it that are not proper to be updated in PdWorker (e.g., readpool's stats).

### TiKV

- [x] Update kvproto to support sending feedback information
    - https://github.com/pingcap/kvproto/pull/1222
- [x] Add HealthController to TiKV, move SlowScore and, SlowTrend to it
    - https://github.com/tikv/tikv/pull/16456
- [x] Send SlowScore as the feedback information to client-go
    - https://github.com/tikv/tikv/pull/16498

### PD

- ~PDClient support getting `StoreStats` field in GetStore. It's already included by the RPC protocol but not returned by the golang PDClient.~

### client-go

- [x] Receive the feedback information and maintain it as part of the store info. 
    - https://github.com/tikv/client-go/pull/1153
- [x] New replica selection implementation that takes the feedback information into consideration.
    - https://github.com/tikv/client-go/pull/1142
- [x] When a request timed out, reload region info quicklier (instead of retrying 10 times as is currently implemented) as the target TiKV might be alive but not serveable.
  - Ref: https://github.com/pingcap/tidb/issues/50432
- [ ] Support actively requesting health feedback form tikv when it's not receiving it for too long (perhaps due to requests are all evicted from that store so that the connection is idle).

Other related works in client-go repo: https://github.com/tikv/client-go/issues/1104

### TiDB

Dependencies of TiDB repo needs to be updated several times, but there isn't any major development task in TiDB repo.
Ref: https://github.com/pingcap/tidb/issues/51412

- [x] https://github.com/pingcap/tidb/pull/51413

### Next step

* Including read path status (e.g. read pool busy)
* TBD.

### PR DESCRIPTION
**Pull request by MyonKeminta**
### What is changed and how it works?

Issue Number: Ref #16297

What's Changed:

```commit-message
Add module health_controller and move SlowScore, SlowTrend, HealthService from PdWorker to it
```

### Review guide

The majority part of this PR is just refactorying.

Code movements in this PR:

* Implementation of `SlowScore`: PdWorker (pd.rs) -> `health_controller::slow_score` (health_controller/src/slow_score.rs)
* Implementation of `SlowTrend`:  components/tikv_util/src/trend.rs -> components/health_controller/src/trend.rs
* Coupling logic between SlowScore and SlowTrend and raftstore:
  * `health_controller::reporters::RaftstoreReporter`
    * `Runner::on_timeout` in pd.rs -> `health_controller::reporters::RaftstoreReporter::tick`
    * `Runner::set_slow_trend_to_store_stats` in pd.rs -> `health_controller::reporter::RaftstoreReporter::update_slow_trend` (except metrics)
  * `health_controller::reporters::SlowTrendStatistics`
  * `health_controller::types`
* Accesses to gRPC's `HealthService` is now wrapped in the `HealthController`.

The design and the structure of the health controller is explained in the documents in health_controller/src/lib.rs.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by LykxSassinator — COMMENTED** [review]
Overall LGTM.

File: components/health_controller/src/reporters.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    pub cause_spike_filter_value_gauge: IntGauge,
+    pub cause_spike_filter_count_gauge: IntGauge,
+    pub cause_l1_gap_gauges: IntGauge,
+    pub cause_l2_gap_gauges: IntGauge,
+
+    pub result_spike_filter_value_gauge: IntGauge,
+    pub result_spike_filter_count_gauge: IntGauge,
+    pub result_l1_gap_gauges: IntGauge,
+    pub result_l2_gap_gauges: IntGauge,
+}
```
Comment (cfzjywxk) [inline components/health_controller/src/reporters.rs:34]: Better to have comments to desribe these fieds or related links about the design.

File: components/health_controller/src/reporters.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+pub struct RaftstoreReporter {
+    health_controller_inner: Arc<HealthControllerInner>,
+    slow_score: SlowScore,
+    slow_trend: SlowTrendStatistics,
+    is_healthy: bool,
+}
+
+impl RaftstoreReporter {
```
Comment (cfzjywxk) [inline components/health_controller/src/reporters.rs:43]: If the read pool information need to be collected, the pattern is:
- Adding a `ReadPoolReporter` reporter holding a `Arc` to the `HealthController`
- Adding related fields inside the  `HealthController`

right?

Besides, a similar question  like above, why holding an `inner` object?

**Review by cfzjywxk — APPROVED** [review]

File: components/health_controller/src/reporters.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/health_controller/src/reporters.rs:34]: IMO, the params related with SlowTrend can directly refer the definition in `components/tikv_utils/src/trend.rs`:

https://github.com/tikv/tikv/blob/5d190fa5c2af88e0cd643aa6368383bb22a83241/components/tikv_util/src/trend.rs#L412-L431

File: components/health_controller/src/reporters.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    pub cause_l2_gap_gauges: IntGauge,
+
+    pub result_spike_filter_value_gauge: IntGauge,
+    pub result_spike_filter_count_gauge: IntGauge,
+    pub result_l1_gap_gauges: IntGauge,
+    pub result_l2_gap_gauges: IntGauge,
+}
+
+pub struct RaftstoreReporter {
+    health_controller_inner: Arc<HealthControllerInner>,
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply components/health_controller/src/reporters.rs:37]: It's just to hide the `Arc` inside and avoid explicitly writing `Arc` when using it, which I think is a common pattern.

File: components/health_controller/src/reporters.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/health_controller/src/reporters.rs:34]: IMO, these params to SlowTrend will be simplified in the later work. Now, just add the reference to the definition in `trend.rs` is good to me.

**Review by ekexium — COMMENTED** [review]
In terms of using these stats, do they tolerate network spikes? Like when network latency becomes unstably high, is it possible to report stats that are stale and out-of-order? Could it possibly cause unexpected behavior when using them (in PD and client)?

pd-worker being stuck could also lead to this. But I assume it very unlikely to happen.

File: components/health_controller/src/reporters.rs (hunk 2 above)
Comment (MyonKeminta) [reply components/health_controller/src/reporters.rs:43]: 1. Yes, but it's not an `Arc` to `HealthController`. Instead, it should be a `Arc<HealthControllerInner>` or a clone of `HealthController`.
2. Separating an "inner" is just to hide the `Arc` inside the `HealthController` and avoid explicitly writing `Arc` when using it, which I think is a common pattern.

**Comment by LykxSassinator** [comment]
> In terms of using these stats, do they tolerate network spikes? Like when network latency becomes unstably high, is it possible to report stats that are stale and out-of-order? Could it possibly cause unexpected behavior when using them (in PD and client)?
> 
> pd-worker being stuck could also lead to this. But I assume it very unlikely to happen.

Yep, it can be detected by `SlowTrendStatistics::record()`. As it describes, it can catch up the normal jitters on network-io, Although current implementation and strategy on the detection of network-io jitters is not sensitive, the optimization on net-io jitters can be tracked in the later work.

File: components/raftstore/src/store/worker/pd.rs [hunk 4]
```diff
@@ -1484,53 +1336,40 @@ where
         self.remote.spawn(f);
     }
 
-    fn set_slow_trend_to_store_stats(
+    fn write_slow_trend_metrics(
```
Comment (LykxSassinator) [inline components/raftstore/src/store/worker/pd.rs:1339]:
```suggestion
    fn flush_slow_trend_metrics(
```

**Review by LykxSassinator — COMMENTED** [review]
Rest LGTM

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**